chips/sam4l/src/eic.rs

```diff
@@ -0,0 +1,375 @@
+            5 => regs.dis.write(Interrupt::INT.val(mask)),
+            6 => regs.dis.write(Interrupt::INT.val(mask)),
+            7 => regs.dis.write(Interrupt::INT.val(mask)),
+            8 => regs.dis.write(Interrupt::INT.val(mask)),
+            _ => debug!("not supported!"),
+        }
+
+        self.line_disable_interrupt(line_num);
+    }
+
+    fn line_configure(
+        &self,
+        line_num: usize,
+        int_mode: hil::eic::InterruptMode,
+        filter: hil::eic::FilterMode,
+        syn_mode: hil::eic::SynchronizationMode,
+    ) {
+        let mask: u32 = 1 << line_num;
+
+        // regs.mode.set(original_mode & !mask);
+
+        let mode_bits = match int_mode {
+            hil::eic::InterruptMode::RisingEdge => 0b00,
+            hil::eic::InterruptMode::FallingEdge => 0b01,
+            hil::eic::InterruptMode::HighLevel => 0b10,
+            hil::eic::InterruptMode::LowLevel => 0b11,
+        };
+
+        self.set_interrupt_mode(mode_bits, mask);
+
+        match filter {
+            hil::eic::FilterMode::FilterEnable => self.line_enable_filter(mask),
+            hil::eic::FilterMode::FilterDisable => self.line_disable_filter(mask),
+        }
+
+        match syn_mode {
+            hil::eic::SynchronizationMode::Synchronous => self.line_disable_asyn(mask),
+            hil::eic::SynchronizationMode::Asynchronous => self.line_enable_asyn(mask),
+        }
+    }
+}
+
+impl<'a> Eic<'a> {
+    pub fn set_interrupt_mode(&self, mode_bits: u8, mask: u32) {
+        let regs: &EicRegisters = &*self.registers;
+        let original_mode: u32 = regs.mode.get();
+        let original_level: u32 = regs.level.get();
+        let original_edge: u32 = regs.edge.get();
+
+        if mode_bits & 0b10 != 0 {
+            regs.mode.set(original_mode & !mask);
+        } else {
+            regs.mode.set(original_mode | mask);
+        }
+
+        if mode_bits & 0b01 != 0 {
+            regs.edge.set(original_edge & !mask); // falling edge
+            regs.level.set(original_level & !mask); // low level
+        } else {
+            regs.edge.set(original_edge | mask); // rising edge
+            regs.level.set(original_level | mask); // high level
+        }
+    }
+
+    const fn new() -> Eic<'a> {
+        Eic {
+            registers: EIC_BASE,
+            enabled: Cell::new(false),
+            callbacks: [
+                OptionalCell::empty(),
+                OptionalCell::empty(),
+                OptionalCell::empty(),
+                OptionalCell::empty(),
+                OptionalCell::empty(),
+                OptionalCell::empty(),
+                OptionalCell::empty(),
+                OptionalCell::empty(),
+                OptionalCell::empty(),
+            ],
+        }
+    }
+
+    pub fn is_enabled(&self) -> bool {
+        self.enabled.get()
+    }
+
+    pub fn set_client(&self, client: &'a hil::eic::Client, line_num: usize) {
+        self.callbacks[line_num].set(client);
```

Comment:
prefer `get()` over direct-indexing to avoid potential array out-of-bounds access

```suggestion
        self.callbacks.get(line_num).map(|c| c.set(client));
```